[WiP] Firefox data source

rubiojr opened the pull request:
Work in progress.

Implements a new Firefox datasource capable of reading its places.sqlite database to import the browser history (page visits).

The implementation currently has a number of issues:

* Firefox (and Firefox based) browser keeps an exclusive lock on the places.sqlite database, and we can't dump or backup it while the browser is open, at least on Linux. To work around that, we copy the database to a temporary directory and import from it. This generally works, but isn't safe, as there's a risk of database corruption when doing the hot copy. Potential alternatives:
  * Ask the user to close the browser while the import happens, which isn't convenient/possible if this is happening regularly in the background.
  * Ignore and retry, as it'll eventually succeed, in the rare case the temporary db copy is corrupted and unreadable
  * Something else, no expert here.
* You need to point Timelinize to the places.sqlite file directly. Pointing it to the Firefox profile directory doesn't seem to work, as it doesn't seem to scan recursively or list all the directory files and pass them to Recognize. I'm probably missing something obvious here.
* Missing tests (will be added)

rubiojr commented:
Screenshot galore!

![Screenshot from 2024-09-11 21-26-04](https://github.com/user-attachments/assets/c7b55ad4-24d8-4ec2-a537-221ca582f991)
![Screenshot from 2024-09-11 21-26-18](https://github.com/user-attachments/assets/a438713c-d74b-4351-9f6a-f6427e18382a)
![Screenshot from 2024-09-11 21-26-12](https://github.com/user-attachments/assets/e6c2a31a-47ba-44e6-ad9f-dbf4bc70dfcc)

mholt commented:
Nice start! Thanks for working on this.

We (I) will need to update the web UI to show bookmarks and web histories in a more useful way than "Item has no content."

About that though... I wonder, should web history have the content be the URL then? Bookmarks are more generalized than web pages and have the semantic notion of "saving" an item. But a page visit... I dunno if there'll be anything else we care to put as the "content". So maybe the URL makes sense to be the content, for web history items. What do you think?

And of course the UI can still show this in some intuitive way, maybe a little SVG background for the URL that looks like the left part of a browser address bar. (Wonder what bookmarks should look like?)

As for the caveats... what if we retry a few times and if it fails to "connect" or has an error reading, we can ask the user to please quit all Firefox windows while importing. Re-importing existing items should be a no-op anyway.

Pointing at the sqlite file directly for now is fine (until we support more stuff from Firefox). Remember that I'm redoing Recognize and Import functions in my local branch and data sources won't have to worry about multiple inputs. So maybe let that be for now.

I haven't looked at the code yet (got a fussy baby at the moment) but I will circle back!

mholt commented:
One other thing I noticed: the visit date doesn't have to be in the metadata, since it's already there in the actual timestamp. :+1:

mholt commented:
I played with a UI to show the page visit on the timeline:

![Screenshot 2024-09-12 at 6 56 47 AM](https://github.com/user-attachments/assets/4e70fb6e-5c86-42af-b6a2-a97fab6964bd)

But then I realized that this would use *way* too much space since each page view needs its own .... that ⬆️.

What I might do for now (probably after this PR is merged) is just show them as a table if there's . title, URL, and description, maybe timestamp as well.

Or, if they're all on the same site, maybe I can group the same site views by this graphic and put a table under it.

I'll be curious to see with my own history what it looks like and work on the UI then.

rubiojr commented:
> About that though... I wonder, should web history have the content be the URL then? Bookmarks are more generalized than web pages and have the semantic notion of "saving" an item. But a page visit... I dunno if there'll be anything else we care to put as the "content". So maybe the URL makes sense to be the content, for web history items. What do you think?

Sounds great to me.

> As for the caveats... what if we retry a few times and if it fails to "connect" or has an error reading, we can ask the user to please quit all Firefox windows while importing. Re-importing existing items should be a no-op anyway.

Yep. Do we have a way to surface importer errors to the users currently?

> Pointing at the sqlite file directly for now is fine (until we support more stuff from Firefox). Remember that I'm redoing Recognize and Import functions in my local branch and data sources won't have to worry about multiple inputs. So maybe let that be for now.

:+1: 

> But then I realized that this would use way too much space since each page view needs its own .... that ⬆️.

Looks gorgeous though :art:

mholt commented:
> Yep. Do we have a way to surface importer errors to the users currently?

Not completely, yet... the UI does actually get all logs, so what we'd probably do is emit logs from a logger with a certain name, and have the UI show users the message. This is still something I'm feeling out however. Will have to come a bit later.

> Looks gorgeous though 🎨

Thanks! It's very simple looking, but will be interesting to implement with HTML/CSS...

mholt commented:
@rubiojr Hello again! I just merged #43 which was a big blocker for new data sources. It would be a good time to rebase. I can help in Discord if you'd like! Note the new `FileImporter` interface.

I _think_ overall things are better this way, but let me know if you run into any snags.

rubiojr commented:
@mholt ✨I'll get back to this next week!

mholt commented on datasources/firefox/firefox.go:
```diff
@@ -0,0 +1,177 @@
+
+// Recognize returns whether the input file is recognized.
+func (Firefox) Recognize(_ context.Context, dirEntry timeline.DirEntry, _ timeline.RecognizeParams) (timeline.Recognition, error) {
+	rec := timeline.Recognition{DirThreshold: .9}
+
+	// we can import directories, but let the import planner figure that out; only recognize files
+	if dirEntry.IsDir() {
+		return rec, nil
+	}
```
This logic means that you can import an entire directory. To explain:

If you have a folder called "Pictures" (the name is irrelevant though), and it's full of .jpg files, the logic you have here is a good way of telling the import planner: "If 90% of the files in the directory match this data source, we'll take the whole folder." And then later, your FileImport() method would be called on the folder, rather than individual files.

I think this data source only imports a single file (places.sqlite), right?

If so, the logic for both Recognize and FileImport can be much simpler.

Just drop these lines, then in FileImport you can remove the WalkDir().

Let me know if I'm assuming wrong :+1:

rubiojr replied:
@mholt no, you're right, although it'd be nice to support importing other kind of files from a firefox profile (such as bookmarks), so importing the whole dir may make sense then, but that'd happen in a future PR.

The changes are still not final though, I was just trying to get it to compile first :wrench:.

rubiojr replied:
@mholt pushed the changes suggested.

mholt reviewed (changes requested):
Looking good so far! Just a little help through the refactor :)

mholt commented:
Hey @rubiojr just curious if you're interested in finishing this up sometime? Is there anything you are waiting on me for?

rubiojr commented:
> Hey @rubiojr just curious if you're interested in finishing this up sometime? Is there anything you are waiting on me for?

@mholt yup, I am, taking a stab today, see if we can get it done this week :pray: .

rubiojr commented:
@mholt rebased and addressed your feedback:

* Simplified import so it only takes a `places.sqlite` file for now
* Added missing data source description
* Set the item content to the visited URL, as discussed in https://github.com/timelinize/timelinize/pull/55#issuecomment-2344639986

Some basic tests still need to be added and I should probably improve logging in case something goes awry while importing the sqlite database, as suggested in https://github.com/timelinize/timelinize/pull/55#issuecomment-2346850702

rubiojr commented:
Somewhat related, https://github.com/timelinize/timelinize/issues/69 will make it a bit harder to import `places.sqlite` in Linux. I have to copy the `places.sqlite` database outside the hidden `~/.mozilla` directory to tests, since I'm unable to navigate to hidden directories using the file importer UI currently.

rubiojr commented:
https://github.com/timelinize/timelinize/pull/71 should fix the lint job failures.

mholt reviewed (commented):
Very nice! How did the new import API work out for you then? Looks like it turned out well. Feeling ready for a final review/merge? At a glance I didn't notice anything major I'd want to change beforehand, if it works...

rubiojr commented:
Worked out great.

I went a bit overboard with the integration tests, and couldn't find much to reuse from existing tests, so let me know how that feels.

Ready for 👀

mholt commented on datasources/firefox/firefox.go:
```diff
@@ -0,0 +1,174 @@
+
+// Recognize returns whether the input file is recognized.
+func (Firefox) Recognize(_ context.Context, dirEntry timeline.DirEntry, _ timeline.RecognizeParams) (timeline.Recognition, error) {
+	if filepath.Base(dirEntry.Name()) == "places.sqlite" {
+		return timeline.Recognition{Confidence: 1}, nil
```
My only thought is, are you sure it's a Firefox DB? I mean, it probably is, but maybe not 100%. I'd set the confidence a little lower, unless you actually open the file and check that it has the needed schema.

Another option would be to look at the full path to see if it's in a Firefox folder (but, this isn't always a surefire solution either, since the file could have been moved/copied). Still, it could raise the confidence a bit if you don't want to open it.

Anyway, this can all be done later, and may never be and still work fine, I know there's plenty of data sources I've written where I am not super thorough with recognizing yet! :sweat_smile:

Anyway, just something to think about, but definitely not a blocker.

rubiojr replied:
> My only thought is, are you sure it's a Firefox DB? I mean, it probably is, but maybe not 100%. I'd set the confidence a little lower, unless you actually open the file and check that it has the needed schema.

I was thinking along those lines today. I could probably open the database and read available tables or the schema to confirm. Checking the path may be tricky given all the Firefox flavors and platforms.

mholt reviewed (approved):
Looks good! Just one thing to consider, but let's go ahead and merge this for now. :smiley: Thank you!!

mholt commented:
(The lint error should be resolved upon merging into main)
